numpy/_core/src/multiarray/multiarraymodule.c

```diff
@@ -1204,9 +1204,24 @@ _pyarray_correlate(PyArrayObject *ap1, PyArrayObject *ap2, int typenum,
      * Need to choose an output array that can hold a sum
      * -- use priority to determine which subtype.
      */
-    ret = new_array_for_sum(ap1, ap2, NULL, 1, &length, typenum, NULL);
-    if (ret == NULL) {
-        return NULL;
+    if (ret != NULL) {
+        if (PyArray_NDIM(ret) != 1) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Output array has wrong dimensionality");
+            return NULL;
+        }
+        if (PyArray_DIMS(ret)[0] != length) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Output array is the wrong shape");
```

Comment:
You could use `PyErr_Format` to print the correct and wrong shapes (not vital, but nice).
Also above, it must be 1-D so we can spell that out (and what it is).